### PR TITLE
i386-more-emu-bugfixes

### PR DESCRIPTION
**Pull request by atlas0fd00m**
bugfix: BSF
bugfix: ROR (emu doesn't set OF when processor does, in the "undefined behavior" realm)
add: e_bits.msb_minus_one() and unittest

### CODE REVIEW

File: envi/archs/i386/emu.py [hunk 1]
```diff
@@ -114,10 +115,68 @@ class BFastCall_Caller(BFastCall):
...
+
+    ret = None
+    while ecx != 0:
+        ret = meth(op)
+        ecx -= 1
+        emu.setRegister(REG_ECX, ecx)
+    return ret
+
+def doRepnzPrefix(emu, meth, op):
+    #FIXME check for opcode family valid to rep
```
Comment (rakuy0) [inline envi/archs/i386/emu.py:135]: We should probably just do this.

File: envi/archs/i386/emu.py [hunk 2]
```diff
@@ -608,10 +651,15 @@ def i_bswap(self, op):
 
     def i_bsr(self, op):
         val = self.getOperValue(op, 1)
-
+        self.setFlag(EFLAGS_AF, False)  # undocumented, but 32-bit VBox on i9 says so
```
Comment (rakuy0) [inline envi/archs/i386/emu.py:654]: Integer now boolean

File: vtrace/envitools.py [hunk 3]
```diff
@@ -4,15 +4,23 @@
 import envi
 import envi.memory as e_memory
 import envi.archs.i386 as e_i386  # FIXME This should NOT have to be here
+import envi.archs.i386.regs as eair
```
[the remark this answers is not in the document]
Comment (rakuy0) [reply vtrace/envitools.py:7]: Honestly, we might want to just merge them into this one and expand the entire lockstepper test.

**Review by rakuy0 — CHANGES_REQUESTED** [review]
Hmm, something seems to have borked the tests?

But still a decent bit of cleanup to go. I love the emulator fixes though.

File: envi/archs/i386/emu.py [hunk 4]
```diff
@@ -1563,15 +1574,14 @@ def i_ror(self, op):
...
-                val = self.getOperValue(op, 0)
+
+            if realcount:
+                print("ror: realcount == 1")
                 cf = self.getFlag(EFLAGS_CF)
-                # FIXME: This may be broke...the manual is kinda flaky here
-                self.setFlag(EFLAGS_OF, e_bits.msb(val, dstSize) ^ (e_bits.msb(val, dstSize) - 1))
-            else:
-                self.setFlag(EFLAGS_OF, 0)
+                self.setFlag(EFLAGS_OF, bool(e_bits.msb(val, dstSize) ^ e_bits.msb_minus_one(val, dstSize)))
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:1647]: i'll argue this one.  bools avoid another shift and act like a zero or a one.

File: envi/archs/i386/regs.py [hunk 5]
```diff
@@ -87,6 +87,14 @@ def getRegOffset(regs, regname):
         ('IF', REG_EFLAGS, 9, 1, 'Interrupt Enable Flag'),
         ('DF', REG_EFLAGS, 10, 1, 'Direction Flag'),
         ('OF', REG_EFLAGS, 11, 1, 'Overflow Flag'),
+        ('OIPL', REG_EFLAGS, 12, 2, 'Overflow Flag'),
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/regs.py:90]: <facepalm>

File: envi/archs/i386/emu.py (hunk 2 above)
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:670]: ok, and i went through and updated the other 15 lines that use boolean instead of integer to set flags.
but i don't like that you're making a big deal about it in this PR... it's literally a bit, it's 1 or 0, True or False.
and these are mostly Visi's doing, i'm just following the pattern that's there.

File: envi/archs/i386/emu.py (hunk 1 above)
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:131]: i just removed it.  i'm not sure how to check it at this point.

File: envi/memory.py [hunk 6]
```diff
@@ -419,7 +405,59 @@ def __init__(self, arch=None):
...
+        '''
+        baseva = None
+        looped = False
+
+        tmpva = suggestaddr
+        while baseva is None:
+            # if we roll into illegal memory, start over at page 2.  skip 0.
+            if tmpva > (1 << (8 * self.imem_psize)):
+                if looped:
+                    raise NoValidFreeMemoryFound(size)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/memory.py:435]: correct.  i'll follow up with a test before pushing to merge.

File: envi/archs/amd64/__init__.py [hunk 7]
```diff
@@ -160,3 +160,66 @@ def i_idiv(self, op):
...
+        cf = (dst >> (src-1)) & 1
+
+        res = e_bits.unsigned(res, dsize)
+
+        self.setFlag(EFLAGS_CF, cf)
+        self.setFlag(EFLAGS_SF, e_bits.is_signed(res, dsize))
+        self.setFlag(EFLAGS_ZF, not res)
+        self.setFlag(EFLAGS_PF, e_bits.is_parity_byte(res))
+        if src == 1:
+            self.setFlag(EFLAGS_OF, False)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/amd64/__init__.py:220]: i need to double-check this one now that your merge is complete.

File: envi/archs/i386/emu.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:1647]: although in this case, i suppose i could  `& 1` instead.

File: envi/archs/amd64/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/amd64/__init__.py:220]: cut this one (and i_sal) from amd64, and fixed i386 one.

File: envi/bits.py [hunk 8]
```diff
@@ -96,9 +96,11 @@ def lsb(value):
     return value & 0x1
 
 def msb(value, size):
-    if value & sign_bits[size]:
-        return 1
-    return 0
+    return bool(value & sign_bits[size])
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/bits.py:99]: because bool and bits are identical, and if/then seems so kludgy when this is sufficient.

File: vivisect/parsers/pe.py [hunk 9]
```diff
@@ -67,7 +67,7 @@ def parseFd(vw, fd, filename=None, baseaddr=None):
     PE.IMAGE_FILE_MACHINE_THUMB: 'thumb16',
 }
 
-defcalls = {
+archcalls = {
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/parsers/pe.py:69]: yes.

File: vtrace/envitools.py [hunk 10]
```diff
@@ -25,36 +33,136 @@ def cmpRegs(emu, trace):
...
+        self.emu = vutil.emuFromTrace(trace)    # this should set up the 
+        self.arch = self.trace.getMeta("Architecture")
+
+    def go(self):
+        count = 0
+        skip_mem_opcodes = skip_mem_opcodes_by_arch.get(self.arch, ())
+
+        self.cmpRegs()
+        while True:
+            print("%4d  Lockstep: 0x%.8x" % (count, self.emu.getProgramCounter()), end='')
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vtrace/envitools.py:54]: ugh.
unfortunately, logging doesn't allow multiple statements to form one logging line.  i'll modify some of them

File: vtrace/envitools.py [hunk 11]
```diff
@@ -25,36 +33,136 @@ def cmpRegs(emu, trace):
...
+        if len(out):
+            raise RegisterException("REGISTER MISMATCH: %s" % '    '.join(out))
 
-    # could use {get,set}MemorySnap if trace inherited from MemoryObject
-    for va, size, perms, fname in trace.getMemoryMaps():
-        try:
-            # So linux maps in a PROT_NONE page for efficient library sharing, so we have to take that into account
-            if (not perms & e_memory.MM_READ):
-                continue
-            if plat == 'linux' and fname in ['[vvar]']:
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vtrace/envitools.py:43]: so, i'm not sure what happened to emuFromTrace.  this looks like a merge fail.

File: vtrace/envitools.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vtrace/envitools.py:7]: that makes sense.  i'm not sure how at the moment.  i don't like LockStepper being in the unittest, but i guess it makes sense on some level.  mine is likely faster (since it doesn't do exhaustive memory comparisons every instruction) but it's also less accurate (for the same reason).

File: vtrace/envitools.py [hunk 12]
```diff
@@ -25,36 +33,136 @@ def cmpRegs(emu, trace):
...
+skip_mem_opcodes_by_arch = {
+        'i386': (e_i386.INS_NOP, e_i386.INS_LEA),
+        'amd64': (e_i386.INS_NOP, e_i386.INS_LEA),
+        }
+
+
+class LockstepEmulator:
+    def __init__(self, trace):
+        self.trace = trace
+        self.emu = vutil.emuFromTrace(trace)    # this should set up the
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vtrace/envitools.py:48]: i believe it somehow got munged during a merge.

File: vtrace/envitools.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vtrace/envitools.py:43]: should be back now.

File: envi/archs/i386/disasm.py [hunk 13]
```diff
@@ -410,7 +410,7 @@ def getOperAddr(self, op, emu=None):
         ret = self.imm
         if emu is not None:
             base, size = emu.getSegmentInfo(op)
-            ret += base
+            ret = e_bits.unsigned(ret + base, emu.imem_psize)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/disasm.py:413]: the addition of negative numbers (eg. 0xfffffffffffffff0) causes an overflow.  it was this or `& e_bits.u_maxes[emu.imem_psize]`

File: envi/archs/i386/disasm.py [hunk 14]
```diff
@@ -524,7 +524,7 @@ def getOperAddr(self, op, emu=None):
         base, size = emu.getSegmentInfo(op)
         ret += base
 
-        return ret + self.disp
+        return e_bits.unsigned(ret + self.disp, emu.imem_psize)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/disasm.py:527]: same

**Comment by atlas0fd00m** [comment]
@rakuy0 i still need to go through and add unittests for the fixed instructions/etc.

File: envi/archs/i386/disasm.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/disasm.py:413]: and yes, effective addresses should always be unsigned ;)

File: vtrace/envitools.py [hunk 15]
```diff
@@ -25,36 +36,134 @@ def cmpRegs(emu, trace):
...
-    '''
-    Produce an envi emulator for this tracer object.
-    '''
-    arch = trace.getMeta('Architecture')
-    plat = trace.getMeta('Platform')
-    amod = envi.getArchModule(arch)
-    emu = amod.getEmulator()
+skip_mem_opcodes_by_arch = {
+        'i386': (e_i386const.INS_NOP, e_i386const.INS_LEA),
+        'amd64': (e_i386const.INS_NOP, e_i386const.INS_LEA),
```
Comment (rakuy0) [inline vtrace/envitools.py:41]: Should we also add in support for ARM/etc?

File: vtrace/envitools.py [hunk 16]
```diff
@@ -25,36 +36,134 @@ def cmpRegs(emu, trace):
...
+                for oper in op.opers:
+                    if skip_mem:
+                        continue
+                    print("(%r = %x)" % (oper.repr(op), oper.getOperValue(op, self.emu)), end='')
+
+                self.trace.stepi()
+                self.emu.stepi()
+                count += 1
+
+                if op.mnem == 'rdtsc':
```
Comment (rakuy0) [inline vtrace/envitools.py:82]: Hmm. It feels like we should copy the concept of a pre/post hook from our emulator class into here, since rdtsc is specifc to intel

File: vtrace/envitools.py [hunk 17]
```diff
@@ -25,36 +36,134 @@ def cmpRegs(emu, trace):
...
+                    self.syncRegsFromTrace()
+
+                print("(eflags: 0x%x)" % self.trace.getRegisterByName('eflags'), end='')
+                print("(eflags: 0x%x)" % self.emu.getRegisterByName('eflags'), end='')
+                self.cmpRegs()
+
+                if not skip_mem:
+                    self.cmpMem(op)
+
+            except RegisterException as msg:
```
Comment (rakuy0) [inline vtrace/envitools.py:92]: Hmm. I feel like we should flesh this portion out. Have a better REPL style workflow, not just for errors, but in general so we can step through each instruction and examine things as they go.

Just a thought. Kind of a big request that we can talk through, but I think it'd be useful.

File: vtrace/envitools.py [hunk 18]
```diff
@@ -1,18 +1,29 @@
...
 
+logger = logging.getLogger(__name__)
+
 class RegisterException(Exception):
     pass
 
+class MemoryException(Exception):
+    pass
+
+class TargetAddrCalcException(Exception):
```
Comment (rakuy0) [inline vtrace/envitools.py:24]: Can you throw these into vtrace/exc.py?

File: vtrace/envitools.py [hunk 19]
```diff
@@ -25,36 +36,134 @@ def cmpRegs(emu, trace):
...
-            print('failed to map: 0x{:x} into emu'.format(va, size))
-            continue
 
-    rsnap = trace.getRegisterContext().getRegisterSnap()
-    emu.setRegisterSnap(rsnap)
+            taddr = oper.getOperAddr(op, self.trace) # use emu here?  it will likely fail the same either way? emu is faster?
+            taddr_emu = oper.getOperAddr(op, self.emu) # use emu here?  it will likely fail the same either way? emu is faster?
+            if taddr != taddr_emu:
+                raise TargetAddrCalcException('TARGET ADDRESS CALCULATED DIFFERENTLY: 0x%x: %r   (%r: emu: 0x%x,  trace: 0x%x)'\
+                        % (op, oper, taddr, taddr_emu))
```
Comment (rakuy0) [inline vtrace/envitools.py:148]: Uhh, you probably want `op.va` instead of op for the first param, since it looks like you're trying to repr an address?

Also, I'm not a fan of just dumping the raw oper. One thing that I just haven't gotten around to (since it's relatively low priority) is changing how opers repr themselves. Right now lot of the operands in like i386 have a `repr` function, but python's builtin stuff calls `__repr__`, so we end up just getting the raw python `<foo class @ anAddressThatIsMeaningless`, so that's something we've gotta fix.

TL;DR: you also probably want `oper.repr(op)` until we fix things.

File: vtrace/util.py [hunk 20]
```diff
@@ -76,3 +78,220 @@ def fireLocalNotifiers(self, event, trace):
...
+    '''
+    Produce an envi emulator for this tracer object.
+
+    If filefmt is None, it will be auto-determined
+
+    If collapse, join adjacent maps
+    If strict, only join maps with the same permissions
+    '''
+    import vivisect
+    import envi.memory as e_memory
```
Comment (rakuy0) [inline vtrace/util.py:144]: We can probably move most of the imports in this function and the other to the top of the file.

File: vtrace/util.py [hunk 21]
```diff
@@ -76,3 +78,220 @@ def fireLocalNotifiers(self, event, trace):
...
+        if trimfname in vw.getFiles(): 
+            continue
+
+        fakemd5 = 'adsfasdfasdfasdfasdf'    
+        # TODO? actually work through all the maps with the same name and add some validish MD5?
+        # but because we're yanking from memory, it won't be the same hash as the binary on disk
+        # since relocations have already be updated and linkage, yadayada
+             
+        vw.addFile(trimfname, va, fakemd5)
+        #FIXME: make sure all of everything is backed by a file...  vw.getFile() should always
```
Comment (rakuy0) [inline vtrace/util.py:213]: Seems like an easy fix we can just add in?

File: vtrace/util.py [hunk 22]
```diff
@@ -76,3 +78,220 @@ def fireLocalNotifiers(self, event, trace):
...
+        # TODO? actually work through all the maps with the same name and add some validish MD5?
+        # but because we're yanking from memory, it won't be the same hash as the binary on disk
+        # since relocations have already be updated and linkage, yadayada
+             
+        vw.addFile(trimfname, va, fakemd5)
+        #FIXME: make sure all of everything is backed by a file...  vw.getFile() should always 
+        #  return a non-None answer
+
+
+        # add segments
```
Comment (rakuy0) [inline vtrace/util.py:217]: Uhhh, yea. Maybe do this?

File: envi/symstore/resolver.py [hunk 23]
```diff
@@ -324,7 +324,9 @@ def getSymList(self):
         """
         Return a list of the symbols which are contained in this resolver.
         """
-        return [self.getSymByName(name) for name in self.symnames]
+        out = [self.getSymByName(name) for name in self.symobjsbyname]
+        out.extend([self.getSymByName(name) for name in self.symnames])
+        return out
```
Comment (rakuy0) [inline envi/symstore/resolver.py:329]: So one question I have is if all the types in the return list are all the same. Well, that and making sure each element in the list is unique.

**Review by rakuy0 — CHANGES_REQUESTED** [review]
I'm liking the changes so far. But still some work to go before we can mainline this I think.

File: envi/archs/i386/emu.py [hunk 24]
```diff
@@ -115,10 +115,64 @@ class BFastCall_Caller(BFastCall):
...
+        ecx -= 1
+        emu.setRegister(REG_ECX, ecx)
+    return ret
+
+
+def doRepSIMDPrefix(emu, meth, op):
+    # TODO
+    raise Exception("doRepSIMDPrefix() not implemented.  Fix and retry.")
+
+class UDException(Exception):
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:146]: absolutely

[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:146]: one catch: UD and BOUND are Intel instructions...

[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:146]: solved the one:  dropped UD and used UnsupportedInstruction which already exists.
BOUND is not so easy.

File: envi/archs/i386/emu.py [hunk 25]
```diff
@@ -115,10 +115,64 @@ class BFastCall_Caller(BFastCall):
 msfastcall_caller = MsFastCall_Caller()
 bfastcall_caller = BFastCall_Caller()
 
+
+def doRepzPrefix(emu, meth, op):
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:119]: sure!

File: vtrace/envitools.py (hunk 19 above)
Comment (atlas0fd00m) [reply vtrace/envitools.py:148]: righto on the op.va!

actually, i want the `<foo class @.....` here, since i want to know what the class is that is a part of the problem.  Sib and Mem can look an awful lot alike.

File: envi/symstore/resolver.py (hunk 23 above)
Comment (atlas0fd00m) [reply envi/symstore/resolver.py:329]: i'm not sure i understand the question.
currently, the way that PE and ELF parsing are written, one populates symobjsbyname and one populates symnames.

Comment (atlas0fd00m) [reply envi/symstore/resolver.py:329]: except i believe all the tiered-naming is done through symobjsbyname?

File: vtrace/util.py (hunk 22 above)
Comment (atlas0fd00m) [reply vtrace/util.py:217]: heh... yep.  sorry, still a WIP.

File: envi/archs/i386/emu.py [hunk 26]
```diff
@@ -1670,7 +1736,7 @@ def i_shr(self, op):
         dst = self.getOperValue(op, 0)
         src = self.getOperValue(op, 1)
 
-        if op.prefixes & PREFIX_REX_W:
+        if op.prefixes & PREFIX_REX_W:  # IA manual states "if 64-bit mode and using REX.W"
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:1739]: ? no?  this is just a comment explaining what's happening.  nothing needs to change here.
i'm trying to document some of the reasons behind some of the esoteric behaviors we're emulating when i think of it.

[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:1739]: i can see that it's not clear enough, tho.

File: envi/archs/i386/emu.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:146]: but done, nonetheless :)

File: envi/archs/i386/emu.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:119]: done.
although the reasoning was that as it was, they didn't require that the emulator exist yet.  the change was a little more than i originally thought, but not terrible.

File: envi/archs/i386/emu.py [hunk 27]
```diff
@@ -617,18 +652,27 @@ def i_bswap(self, op):
 
     def i_bsr(self, op):
         val = self.getOperValue(op, 1)
+        eflags = self.getRegister(REG_EFLAGS)
+        eflags &= 0x00044602        # undocumented, empirical from i9 - do we want to move to main emulator?
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:656]: i'd rather not... that would add significant performance cost to a relatively simple instruction.  meanwhile, it's entirely possible we'll want to do this to a lot more instructions, and i don't want to incur the costs of so many extra calculations this and other instructions.
this is &= which means getting and conditionally setting each of 32 flags.  
and i'm not entirely sure why it's doing it, this is what's seen on a real processor.

it's possible that the emulator itself should be doing some of this in between instructions.

but i'm happy to document what is being masked or allowed.

File: vtrace/util.py (hunk 21 above)
Comment (atlas0fd00m) [reply vtrace/util.py]: that was more a reminder for me to make sure it was working.

File: vtrace/platforms/win32.py [hunk 28]
```diff
@@ -1617,7 +1617,7 @@ def platformProcessEvent(self, event):
             self.setMeta('IsWow64', self._is_wow64)
 
             self.fireNotifiers(vtrace.NOTIFY_ATTACH)
-            self.addLibraryBase(ImageName, baseaddr)
+            self.addLibraryBase(ImageName, baseaddr, always=True)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vtrace/platforms/win32.py:1620]: i was pretty convinced that this was necessary for my approach to fixing the Vtrace naming.  i'll have to go back and see why i did this.  this was a carryover from the Vtrace PR and i'm not sure why i allowed it to commit to this one.

can you explain what happens that is bad?

File: vtrace/util.py (hunk 20 above)
Comment (atlas0fd00m) [reply vtrace/util.py]: done

File: envi/symstore/resolver.py [hunk 29]
```diff
@@ -201,7 +201,7 @@ def delSymbol(self, sym):
             subres = self.symnames.get(sym.fname)
 
         # Potentially del it from the sub resolver's namespace
-        if subres is not None:
+        if subres is not None and not isinstance(subres, tuple):
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/symstore/resolver.py:204]: because we store sub-resolvers in symnames, not just symbols.  without this, we're ending up with trying to use a tuple (symbol) as a sub-resolver.

this may be a great start to symbol management, but it requires a bit of love, more than can be done in either of these two PR's i'm working on.

File: envi/tests/test_memory.py [hunk 30]
```diff
@@ -19,3 +20,39 @@ def test_envi_memory_cache(self):
...
+
+        newmapva3 = mem.allocateMemory(1024000, name='test3', fill=b'@')
+        self.assertNotEqual(newmapva3, newmapva2)
+        self.assertEqual(b'@@@@', mem.readMemory(newmapva3, 4))
+
+        newmapva4 = mem.allocateMemory(1024000, suggestaddr=0x41410000, name='test4', fill=b'@')
+        self.assertNotEqual(0x41410000, newmapva4)
+
+        newmapva5 = mem.allocateMemory(1024000, perms=e_mem.MM_READ, suggestaddr=0x41410000, name='test4', fill=b'@')
+        #self.assertRaises(e_exc.SegmentationViolation, mem.writeMemory(newmapva5, "foobarbaz"))  # throws the correct exception, but still somehow fails.
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/tests/test_memory.py]: nm.  i needed to use `with self.assertRaises(exc): function()`

File: envi/memory.py [hunk 31]
```diff
@@ -419,7 +406,59 @@ def __init__(self, arch=None):
...
+
+        tmpva = suggestaddr
+        while baseva is None:
+            # if we roll into illegal memory, start over at page 2.  skip 0.
+            if tmpva > (1 << (8 * self.imem_psize)):
+                if looped:
+                    raise e_exc.NoValidFreeMemoryFound(size)
+
+                looped = True
+                tmpva = 0x1000
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/memory.py:439]: don't want to be based at 0.  call it personal preference, but bad things happen when you base things at 0.  it's a 4k boundary above 0.

File: envi/tests/test_memory.py [hunk 32]
```diff
@@ -19,3 +20,39 @@ def test_envi_memory_cache(self):
...
+        mem = e_mem.MemoryObject()
+        mem.addMemoryMap(0x41410000, e_mem.MM_RWX, 'test', b'\0'*1024)
+        self.assertEqual(mem.readMemory(0x41410041, 4), b'\0\0\0\0')
+        mem.writeMemory(0x41410041, b'foo')
+        self.assertEqual(mem.readMemory(0x41410041, 4), b'foo\0')
+
+        mem.delMemoryMap(0x41410000)
+
+        # these next two tests *must* fail.  not failing would be the failure.
+        self.assertRaises(e_exc.SegmentationViolation, mem.readMemory, 0x41410041, 4)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/tests/test_memory.py:34]: yup!  thanks. that helped my other problem

File: vtrace/envitools.py (hunk 17 above)
Comment (atlas0fd00m) [reply vtrace/envitools.py:92]: i'd love the conversation.  i like mine for some things (interactive), but i could be convinced that yours is better.

File: envi/archs/i386/emu.py [hunk 33]
```diff
@@ -1554,7 +1564,8 @@ def i_rol(self, op):
     def i_ror(self, op):
         dstSize = op.opers[0].tsize
         count = self.getOperValue(op, 1)
-        tempCount = shiftMask(count, dstSize)
+        realcount = shiftMask(count, dstSize)
+        tempCount = realcount
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:1620]: isn't that enough?

File: envi/archs/i386/opconst.py [hunk 34]
```diff
@@ -65,7 +65,7 @@
 
 OPERSIZE = {
     0: (2, 4, 8),           # We will only end up here on regs embedded in opcodes
-    OPTYPE_a: (2, 4, 4),
+    OPTYPE_a: (4, 8, 8),
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/opconst.py:68]: yup.  one-word is 16-bits.

File: envi/archs/i386/emu.py [hunk 35]
```diff
@@ -672,7 +679,7 @@ def i_cli(self, op):
     def i_cmc(self, op):
         # set the CF flag to its complement
         val = self.getFlag(EFLAGS_CF)
-        self.setFlag(EFLAGS_CF, ~val)
+        self.setFlag(EFLAGS_CF, not val)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/archs/i386/emu.py:731]: don't know.  it just doesn't work the same.

File: envi/const.py [hunk 36]
```diff
@@ -11,3 +11,34 @@
 
 ENDIAN_LSB = 0
 ENDIAN_MSB = 1
+
+# Memory Map Permission Flags
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/const.py:15]: will do.

File: vtrace/envitools.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vtrace/envitools.py:7]: agreed.

File: vivisect/analysis/i386/calling.py [hunk 37]
```diff
@@ -52,6 +53,20 @@ def prehook(self, emu, op, starteip):
...
+        if op.opcode == e_i386const.INS_LEA:    # x86 only
+            i = 1
+            o = op.opers[i]
+            discrete = o.isDiscrete()
+            operva = o.getOperAddr(op, emu)
+            # keep track of the max here, but save it for later too...
+            stackoff = emu.getStackOffset(operva)
+            if stackoff and stackoff >= 0:
+                self.stackmax = max(self.stackmax, stackoff)
+                self.stackargs[stackoff] = True
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vivisect/analysis/i386/calling.py:65]: ?
if you can find a more elegant way to do this, i'm open (truly).
this makes lea-accesses of local variables work.  they did not before.
from my point of view, there's zero room for skepticism on that part.

i understand where you're coming from.  it's just a big deal and a huge win to get local var tracking on LEA instructions

File: vtrace/envitools.py (hunk 15 above)
Comment (atlas0fd00m) [reply vtrace/envitools.py:41]: i don't know of any offhand for other architectures.  we can certainly add them as we come across them.

File: vtrace/envitools.py (hunk 16 above)
Comment (atlas0fd00m) [reply vtrace/envitools.py:82]: agreed.  or at least a architecture-specific handler system.  rdtsc reads from the system clock, which we don't do.  so this triggers a register-resync.

File: vtrace/envitools.py (hunk 17 above)
Comment (atlas0fd00m) [reply vtrace/envitools.py:92]: i love the REPL approach.  let's make this a different PR.

File: vtrace/envitools.py (hunk 18 above)
Comment (atlas0fd00m) [reply vtrace/envitools.py:24]: done

File: envi/tests/test_memory.py [hunk 38]
```diff
@@ -19,3 +20,39 @@ def test_envi_memory_cache(self):
...
+        self.assertRaises(e_exc.SegmentationViolation, mem.writeMemory, 0x41410041, b'foobar')
+
+    def test_allocator(self):
+        mem = e_mem.MemoryObject()
+        mem.addMemoryMap(0x41410000, e_mem.MM_RWX, 'test', b'\0'*1024)
+        newmapva1 = mem.findFreeMemoryBlock(1024000)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/tests/test_memory.py:41]: not at all easily...

[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply envi/tests/test_memory.py:41]: done anyway. :)

File: vtrace/util.py [hunk 39]
```diff
@@ -76,3 +78,220 @@ def fireLocalNotifiers(self, event, trace):
...
+        psize = trace.getPointerSize()
+        # capture PEB and TIB
+        peb = trace.getMeta('PEB')
+        tebs = dict(trace.win32threads)
+        # test and make sure:
+        for threadid, teb in tebs.items():
+            assert(trace.readMemoryPtr(teb + (psize * 12)) == peb)
+            assert(trace.readMemoryPtr(teb + (psize * 9)) == threadid)
+
+        emu.setMeta('PEB', peb)
```
[the remark this answers is not in the document]
Comment (atlas0fd00m) [reply vtrace/util.py]: i have custom consumers of it. :)

**Review by rakuy0 — APPROVED** [review]
